### DISCUSSION TITLE
Help with writing a query using multiple 'as' using with cte

### DISCUSSION THREAD

**Discussion by isidpllc** [linq2db/linq2db#4579] [Q&A]
Please help me write this query using linq2db

```
WITH
	CTE AS (
		SELECT
			ID
		FROM
			TABLE1
		WHERE
			* * *
	),
	UPDATEEXISTS AS (
		UPDATE TABLE2
		SET
			FIELD1 = @SOMEVALUE1
		FROM
			CTE
		WHERE
			TABLE2.FIELD2 = CTE.ID
	)
INSERT INTO
	TABLE2 (ID, FIELD1, FIELD2)
SELECT
	GEN_RANDOM_UUID (),
	@SOMEVALUE1 CTE.ID
FROM
	CTE
	LEFT JOIN TABLE2 ON TABLE2.FIELD2 = CTE.ID
WHERE
	TABLE2.ID IS NULL
```

I understand how to do this in two separate requests (for example, first a request for update, and then for insert).
Like that

```
var query = ***.AsCte();

await _context.GetTable<Table2>()
	.LeftJoin(query, (table2, table1) => table2.Field2 == table1.Id, (table2, table1) => new { Table2 = table2, Object = table1 })
	.Set(x => x.Table2.Field1, someValue)
	.UpdateAsync();
```

But I don’t understand how to do this in one request.

**Comment by sdanyliv** [comment]
`UPDATE` in CTE is not supported. Which database do you use and what are you trying to achive at the end? We do not understand from the SQL what is needed.

**Reply by isidpllc** [reply]
@sdanyliv , PostgreSQL

**Reply by sdanyliv** [reply]
I have updated question. We do not understand why it is needed. linq2db supports `MERGE`, maybe it will be enough.

**Comment by MaceWindu** [comment]
If you use Pgsql 15+ you can use merge. Should look something like that
```cs
db.Table2
  .Merge()
  .Using(db.Table1.Where(...))
  .On((t2, t1) => t2.Field2 == t1.ID)
  .InsertWhenNotMatched(...)
  .UpdateWhenMatched(...)
  .Merge();
```

**Reply by isidpllc** [reply]
@MaceWindu , The method 'InsertWhenNotMatched' requires the same <Target, Target> types.
But thanks for the tip, I’ll do some more research in this direction.
Postgres15 is not yet available in our project, so we will have to make two separate requests.
